dvc/repo/get.py

```diff
@@ -29,32 +30,41 @@ def get(url, path, out=None, rev=None):
     # reflink and/or hardlink. Not using tempfile.TemporaryDirectory
     # because it will create a symlink to tmpfs, which defeats the purpose
     # and won't work with reflink/hardlink.
-    dpath = os.path.dirname(os.path.abspath(out))
+    abspath = os.path.abspath(out)
+    dpath = os.path.dirname(abspath)
     tmp_dir = os.path.join(dpath, "." + str(shortuuid.uuid()))
     try:
         with external_repo(cache_dir=tmp_dir, url=url, rev=rev) as repo:
-            # Note: we need to replace state, because in case of getting DVC
-            # dependency on CIFS or NFS filesystems, sqlite-based state
-            # will be unable to obtain lock
-            repo.state = StateNoop()
+            tmp_repo_path = os.path.join(repo.root_dir, path)
```

Comment:
After a discussion with @efiop, we noticed that, if `path` is absolute,
we will be able to get any file available on the system. In mentioned case, `join` will just return `path`,
allowing us to copy literally anything.

Solution to this could be to check for `dvc`-controlled file first, and only then, if it does not exist,
try to find git controlled file. Additionaly,  we would need to forbid absolute paths for files that are not controlled by `dvc`. 

Also, we already had the same problem, when [writing get for api](https://github.com/iterative/dvc/pull/2566/files#diff-16d20341660af559adf7d9e7306c760d)